File: src/main/java/net/hectus/neobb/game/util/EffectManager.java
package net.hectus.neobb.game.util;

import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.turn.Turn;
import net.hectus.neobb.turn.default_game.block.BlockTurn;
import net.hectus.neobb.turn.default_game.mob.MobTurn;
import net.hectus.neobb.turn.default_game.structure.StructureTurn;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Shulker;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scoreboard.Team;
import org.jetbrains.annotations.NotNull;

public class EffectManager {
    private LivingEntity highlight;

    public void applyEffects(@NotNull Turn<?> turn) {
        NamedTextColor outlineColor = turn.player().cosmeticOutline();

        if (turn instanceof BlockTurn blockTurn) highlightBlock(blockTurn.data(), outlineColor);
        if (turn instanceof StructureTurn structureTurn) highlightBlock(structureTurn.data().lastBlock(), outlineColor);
        if (turn instanceof MobTurn<?> mobTurn) applyHighlight(mobTurn.data(), outlineColor);

        spawnParticle(turn.location(), turn.player());
    }

    public void spawnParticle(Location loc, @NotNull NeoPlayer player) {
        player.cosmeticPlaceParticle().spawn(loc);
    }

    public void highlightBlock(@NotNull Block block, NamedTextColor outlineColor) {
        if (block.getType().name().contains("GLASS")) return;

        block.getWorld().spawn(block.getLocation(), Shulker.class, shulker -> {
            shulker.setAI(false);
            shulker.setInvisible(true);
            shulker.setInvulnerable(true);
            applyHighlight(shulker, outlineColor);
        });
    }

    public void applyHighlight(@NotNull LivingEntity entity, NamedTextColor outlineColor) {
        clearHighlight();
        highlight = entity;

        entity.addPotionEffect(new PotionEffect(PotionEffectType.GLOWING, -1, 0, false, false));

        Team team = Bukkit.getScoreboardManager().getMainScoreboard().registerNewTeam("highlight-" + entity.getUniqueId().getMostSignificantBits());
        team.color(outlineColor);
        team.addEntity(entity);
    }

    public void clearHighlight() {
        if (highlight != null) {
            if (highlight instanceof Shulker) {
                highlight.remove();
            } else {
                highlight.removePotionEffect(PotionEffectType.GLOWING);
            }
        }
    }
}

File: src/main/kotlin/net/hectus/neobb/cosmetic/EffectManager.kt
package net.hectus.neobb.cosmetic

import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.turn.Turn
import net.hectus.neobb.turn.default_game.block.BlockTurn
import net.hectus.neobb.turn.default_game.mob.MobTurn
import net.hectus.neobb.turn.default_game.structure.StructureTurn
import net.hectus.neobb.util.Constants
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.Bukkit
import org.bukkit.Color
import org.bukkit.Location
import org.bukkit.block.Block
import org.bukkit.entity.BlockDisplay
import org.bukkit.entity.Display
import org.bukkit.entity.Entity
import org.bukkit.scoreboard.Team
import org.bukkit.util.Transformation
import org.joml.AxisAngle4f
import org.joml.Vector3f

class EffectManager {
    private var highlight: Pair<Entity, Team?>? = null

    fun applyEffects(turn: Turn<*>) {
        if (turn.player == null) return

        turn.player.playSound(turn.player.databaseInfo.placeSound)

        val outlineColor = turn.player.databaseInfo.outline
        if (turn is BlockTurn) highlightBlock(turn.data!!, outlineColor)
        if (turn is StructureTurn) highlightBlock(turn.data!!.lastBlock, outlineColor)
        if (turn is MobTurn<*>) applyHighlight(turn.data!!, outlineColor)

        spawnParticle(turn.location(), turn.player)
    }

    fun spawnParticle(location: Location, player: NeoPlayer) {
        player.databaseInfo.placeParticle.spawn(location)
    }

    fun highlightBlock(block: Block, outlineColor: NamedTextColor) {
        clearHighlight()

        val display = block.world.spawn(block.location, BlockDisplay::class.java)
        display.block = block.blockData.clone()
        display.transformation = Transformation(
            Vector3f(0.0f),
            AxisAngle4f(0.0f, 0.0f, 0.0f, 0.0f),
            Vector3f(Constants.HIGHLIGHT_SCALE),
            AxisAngle4f(0.0f, 0.0f, 0.0f, 0.0f)
        )
        display.brightness = Display.Brightness(block.lightFromBlocks.toInt(), block.lightFromSky.toInt())

        display.glowColorOverride = Color.fromRGB(outlineColor.value())
        display.isGlowing = true

        highlight = display to null
    }

    fun applyHighlight(entity: Entity, outlineColor: NamedTextColor) {
        clearHighlight()

        val team = Bukkit.getScoreboardManager().mainScoreboard.registerNewTeam("highlight-${entity.uniqueId.mostSignificantBits}")
        team.color(outlineColor)
        team.addEntity(entity)
        highlight = entity to team
    }

    fun clearHighlight() {
        val highlight = highlight ?: return
        when (highlight.first) {
            is BlockDisplay -> highlight.first.remove()
            else -> highlight.first.isGlowing = false
        }
        highlight.second?.unregister()
    }
}
